### ISSUE TITLE
Multiple

### ISSUE THREAD

**Issue by BehaviorEconomist** [BehaviorEconomist/Forecasting#2]
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

# Import 2023 Observations by Month for time-series
April = pd.read_excel('Issue Summary Monthly 04_30_2023.xlsx')
May = pd.read_excel('Issue Summary Monthly 05_31_2023.xlsx')
June = pd.read_excel('Issue Summary Monthly 06_30_2023.xlsx')
July = pd.read_excel('Issue Summary Monthly 07_31_2023.xlsx')
August = pd.read_excel('Issue Summary Monthly 08_31_2023.xlsx')
September = pd.read_excel('Issue Summary Monthly 09_30_2023.xlsx')
October = pd.read_excel('Issue Summary Monthly 10_31_2023.xlsx')

# Define the categories and months
categories = ['Bank', 'Weather', 'Risk']
months = ['April', 'May', 'June', 'July', 'August', 'September', 'October']

# Convert month names to a numeric format for linear regression
month_to_num = {name: num for num, name in enumerate(months, start=1)}
num_months = [month_to_num[month] for month in months]
forecast_months_num = [month_to_num['October'] + i for i in range(1, 4)]
forecast_months = ['November', 'December', 'January']

# Initialize a dictionary to store DataFrames for each category
category_data = {}

for category in categories:
    # Filter and count unique ISSUE_IDs for each month
    monthly_counts = []
    for month_df in [April, May, June, July, August, September, October]:
        filtered_df = month_df[month_df['OWN RSRC Level 04 TXT'] == category]
        monthly_counts.append(filtered_df['ISSUE_ID'].nunique())

    # Linear Regression Model
    X = np.array(num_months).reshape(-1, 1)
    y = np.array(monthly_counts)
    model = LinearRegression()
    model.fit(X, y)

    # Predict for future months
    forecast_counts = model.predict(np.array(forecast_months_num).reshape(-1, 1))

    # Create and store DataFrames for historical and forecasted values
    historical_df = pd.DataFrame({'Month': months, 'Issue_Count': monthly_counts})
    forecast_df = pd.DataFrame({'Month': forecast_months, 'Issue_Count': forecast_counts})
    combined_df = pd.concat([historical_df, forecast_df]).reset_index(drop=True)
    category_data[category] = combined_df

    # Separate Visualization for each category
    plt.figure(figsize=(12, 6))
    plt.plot(combined_df['Month'], combined_df['Issue_Count'], marker='o')
    plt.title(f'Monthly Issues Count with Forecast for {category}')
    plt.xlabel('Month')
    plt.ylabel('Number of Issues')
    plt.xticks(rotation=45)
    plt.grid(True)
    plt.show()

# Combined Visualization for all categories
plt.figure(figsize=(12, 6))
for category, data in category_data.items():
    plt.plot(data['Month'], data['Issue_Count'], marker='o', label=category)

plt.title('Combined Monthly Issues Count with Forecast')
plt.xlabel('Month')
plt.ylabel('Number of Issues')
plt.xticks(rotation=45)
plt.grid(True)
plt.legend()
plt.show()
